### ISSUE TITLE
Bittrex - Crash when cancelling an unfilled buy order.

### ISSUE THREAD

**Issue by mostlyvirtual** [freqtrade/freqtrade#3130]
- Operating system: Ubuntu 18.04.4
- Python Version: 3.6.9 (python -V)
- CCXT version: 1.24.83 (pip freeze | grep ccxt)
- Branch: master 2020.3
- Last Commit ID: 9abfdaaaa240dcab2fbd105982ed6eed025d8936
  
## Describe the problem:

*When the bot tries to place an order that doesn't get filled, and hits the time limit, it tries to cancel that order and crashes. 
It happened 2 times thus far.*

### Steps to reproduce:

  1. Place an order on Bittrex that doesn't get filled - I was using the HBAR/USD pairs in this example.
  2. Hit the unfilledtimeout limit.
  3. Crash.

### Relevant code exceptions or logs:

  ```
2020-04-01 18:20:37,872 - freqtrade.freqtradebot - INFO - Buy order cancelled due to timeout for Trade(id=23, pair=HBAR/USD, amount=1286.13227870, open_rate=0.03110100, open_since=2020-04-01 16:10:35).
2020-04-01 18:20:37,873 - freqtrade.commands.trade_commands - INFO - worker found ... calling exit
2020-04-01 18:20:37,873 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'process died'}
2020-04-01 18:20:37,937 - freqtrade.freqtradebot - INFO - Cleaning up modules ...
2020-04-01 18:20:37,938 - freqtrade.rpc.rpc_manager - INFO - Cleaning up rpc modules ...
2020-04-01 18:21:05,174 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/root/ft-bt-l-hbar-usd-Strategy005/freqtrade/main.py", line 36, in main
    return_code = args['func'](args)
  File "/root/ft-bt-l-hbar-usd-Strategy005/freqtrade/commands/trade_commands.py", line 20, in start_trading
    worker.run()
  File "/root/ft-bt-l-hbar-usd-Strategy005/freqtrade/worker.py", line 66, in run
    state = self._worker(old_state=state)
  File "/root/ft-bt-l-hbar-usd-Strategy005/freqtrade/worker.py", line 104, in _worker
    self._throttle(func=self._process_running, throttle_secs=self._throttle_secs)
  File "/root/ft-bt-l-hbar-usd-Strategy005/freqtrade/worker.py", line 125, in _throttle
    result = func(*args, **kwargs)
  File "/root/ft-bt-l-hbar-usd-Strategy005/freqtrade/worker.py", line 139, in _process_running
    self.freqtrade.process()
  File "/root/ft-bt-l-hbar-usd-Strategy005/freqtrade/freqtradebot.py", line 158, in process
    self.check_handle_timedout()
  File "/root/ft-bt-l-hbar-usd-Strategy005/freqtrade/freqtradebot.py", line 876, in check_handle_timedout
    self.handle_timedout_limit_buy(trade, order)
  File "/root/ft-bt-l-hbar-usd-Strategy005/freqtrade/freqtradebot.py", line 917, in handle_timedout_limit_buy
    trade.amount = order['amount'] - corder.get('remaining', order['remaining'])
TypeError: unsupported operand type(s) for -: 'float' and 'NoneType'
  ```

**Comment by xmatthias** [comment]
I think i've found the problem - seems like 'remaining' can be None - but the key would still be there - should be fixed in #3140.

**Comment by mostlyvirtual** [comment]
I've switched to that branch and have begun testing.
version='develop-68a5e0c5'

I will let you know.

**Comment by xmatthias** [comment]
are you sure this worked?

`68a5e0c5` sounds very much like develop to me ... (log message "Merge pull request #3138 from orkblutt/develop")...

should be more like `develop-c03f637f` ...

**Comment by mostlyvirtual** [comment]
Yes, it seemed to have worked:

# git branch
* buy_order_timeout_logging
  develop

---

Also, I had my first canceled order today, and I confirm that it was correctly canceled with no crash.

Thanks again.

**Comment by mostlyvirtual** [comment]
Not sure if this is related, but the /balance seems to be incorrect now.

It still shows the HBAR trade in the balance as active - the one it cancelled.

And it doesn't show the new trade it made this morning on BTC.

**Comment by xmatthias** [comment]
Thanks for confirming :+1: 

I strongly doubt that it's related at all to the balance problem 

`/balance` pulls the balance from the exchange (assuming live/production mode)- so best verify on the exchange if you really do not own the asset (i'm not using bittrex myself anymore - but `/balance` is accurate for binance, ftx and kraken for me)  ...

Also - keep an eye on the logs - /balance does cache the balance for up to an hour if no "bot-action" happens - to avoid getting you blocked by frequently calling `/balance`.

However, i'd like not to merge issues - so if you encounter a "mismatch" - which doesn't go away within an hour (or a bot restart) - best open a new issue.

**Comment by hroff-1902** [comment]
@mostlyvirtual the fix was merged into develop, pls verify

**Comment by mostlyvirtual** [comment]
@hroff-1902 sorry - I didn't have a chance to test this yet - I see that @xmatthias said it was merged into 2020.4, so can I be on master and test this, or do I have to stay on dev?

**Comment by xmatthias** [comment]
i didn't mention it intentionally - but i think i had it as part of a commit message.

either way - 2020.4 does include this fix.

**Comment by mostlyvirtual** [comment]
@xmatthias thanks - I've updated a couple of bots that were failing on Bittrex - I'll update this thread in a few days after they ran some cancellations.

**Comment by mostlyvirtual** [comment]
I think the issue might still be here, though I do want to say that it cancelled *some* of the orders before failing - originally it wouldn't' be able to cancel any orders.

```
2020-05-12 00:59:21,186 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=15, pair=HBAR/ETH, amount=2230.89793642, open_rate=0.00017930, open_since=2020-05-11 22:29:22)
2020-05-12 00:59:26,242 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=15, pair=HBAR/ETH, amount=2230.89793642, open_rate=0.00017930, open_since=2020-05-11 22:29:22)
2020-05-12 00:59:27,756 - freqtrade.freqtradebot - INFO - Buy order cancelled due to timeout for Trade(id=15, pair=HBAR/ETH, amount=2230.89793642, open_rate=0.00017930, open_since=2020-05-11 22:29:22).
2020-05-12 00:59:27,757 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=15, pair=HBAR/ETH, amount=2155.33386951, open_rate=0.00017930, open_since=2020-05-11 22:29:22)
2020-05-12 00:59:29,300 - freqtrade.freqtradebot - INFO - Applying fee on amount for Trade(id=15, pair=HBAR/ETH, amount=2155.33386951, open_rate=0.00017930, open_since=2020-05-11 22:29:22) failed: myTrade-Dict empty found
2020-05-12 00:59:29,301 - freqtrade.commands.trade_commands - INFO - worker found ... calling exit
2020-05-12 00:59:29,301 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'process died'}
2020-05-12 00:59:29,455 - freqtrade.freqtradebot - INFO - Cleaning up modules ...
2020-05-12 00:59:29,455 - freqtrade.rpc.rpc_manager - INFO - Cleaning up rpc modules ...
2020-05-12 00:59:30,526 - freqtrade - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/main.py", line 36, in main
    return_code = args['func'](args)
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/commands/trade_commands.py", line 20, in start_trading
    worker.run()
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/worker.py", line 66, in run
    state = self._worker(old_state=state)
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/worker.py", line 104, in _worker
    self._throttle(func=self._process_running, throttle_secs=self._throttle_secs)
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/worker.py", line 125, in _throttle
    result = func(*args, **kwargs)
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/worker.py", line 139, in _process_running
    self.freqtrade.process()
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/freqtradebot.py", line 150, in process
    self.check_handle_timedout()
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/freqtradebot.py", line 879, in check_handle_timedout
    self.handle_timedout_limit_buy(trade, order)
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/freqtradebot.py", line 927, in handle_timedout_limit_buy
    self.update_trade_state(trade, corder, trade.amount)
  File "/root/ft-bt-l-hbar-eth-ADXM/freqtrade/freqtradebot.py", line 1144, in update_trade_state
    if not isclose(order['amount'], new_amount, abs_tol=constants.MATH_CLOSE_PREC):
TypeError: must be real number, not NoneType
```

After looking at the block of code that's failing, at line 1144, I see this has been added as a try/except for a "binance-fix".
So the issue right now might be out of scope to the original bug that was filed? Which would explain why it can close the buy order sometimes, but failed in this instance, which might be an edge case?
Should I open a new issue for this?

**Comment by xmatthias** [comment]
the only explanation for this new error would be if get_order returns a dict with the amount key empty.

Now the problem is - if that were the case for Bittrex - it would not work at all - as that is a required field, and is used in multiple other instances as well.


But i wonder - which version are you running when encountering the above error?

checking the current develop, line 1144 (https://github.com/freqtrade/freqtrade/blob/8e4ffea52b72d88f03bcf41d994dfa37b81a88b3/freqtrade/freqtradebot.py#L1144) is not something that could raise this error.

**Comment by datanerdscott** [comment]
I started receiving this error again on Bittrex starting on August 5th, which has been fixed since I updated to 2020.04 back in May.  It too is intermittent and does not always fail. On Aug 5th I was running 2020.06 and have since updated to 2020.07 with no luck.  Should I open a new issue for this or is it related to the fix implemented in 2020.04?  See below for recent error message and let me know if any additional info would be helpful.  Thanks!

2020-08-17 14:14:16,178 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=81, pair=HBAR/USD, amount=39337.37760912, open_rate=0.05084223, open_since=2020-08-10 21:03:09)
2020-08-17 14:14:17,668 - freqtrade.wallets - INFO - Wallets synced.
2020-08-17 14:14:19,208 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=113, pair=DGB/USD, amount=55187.63796909, open_rate=0.03624000, open_since=2020-08-17 05:15:07)
2020-08-17 14:14:20,743 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=118, pair=ETC/USD, amount=270.25349337, open_rate=7.40045938, open_since=2020-08-17 14:11:12)
2020-08-17 14:14:22,284 - freqtrade.freqtradebot - INFO - Buy order cancelled due to timeout for Trade(id=118, pair=ETC/USD, amount=270.25349337, open_rate=7.40045938, open_since=2020-08-17 14:11:12).
2020-08-17 14:14:22,284 - freqtrade.freqtradebot - INFO - Found open order for Trade(id=118, pair=ETC/USD, amount=39.86523463, open_rate=7.40045938, open_since=2020-08-17 14:11:12)2020-08-17 14:14:23,679 - freqtrade.freqtradebot - INFO - Applying fee on amount for Trade(id=118, pair=ETC/USD, amount=39.86523463, open_rate=7.40045938, open_since=2020-08-17 14:11:12) failed: myTrade-Dict empty found
2020-08-17 14:14:23,680 - freqtrade.commands.trade_commands - ERROR - must be real number, not NoneType
2020-08-17 14:14:23,680 - freqtrade.commands.trade_commands - ERROR - Fatal exception!
Traceback (most recent call last):
  File "/home/ubuntu/freqtrade/freqtrade/commands/trade_commands.py", line 20, in start_trading
    worker.run()
  File "/home/ubuntu/freqtrade/freqtrade/worker.py", line 73, in run
    state = self._worker(old_state=state)
  File "/home/ubuntu/freqtrade/freqtrade/worker.py", line 110, in _worker
    self._throttle(func=self._process_running, throttle_secs=self._throttle_secs)
  File "/home/ubuntu/freqtrade/freqtrade/worker.py", line 131, in _throttle
    result = func(*args, **kwargs)
  File "/home/ubuntu/freqtrade/freqtrade/worker.py", line 144, in _process_running
    self.freqtrade.process()
  File "/home/ubuntu/freqtrade/freqtrade/freqtradebot.py", line 162, in process
    self.check_handle_timedout()
  File "/home/ubuntu/freqtrade/freqtrade/freqtradebot.py", line 937, in check_handle_timedout
    self.handle_cancel_buy(trade, order, constants.CANCEL_REASON['TIMEOUT'])
  File "/home/ubuntu/freqtrade/freqtrade/freqtradebot.py", line 1003, in handle_cancel_buy
    self.update_trade_state(trade, corder, trade.amount)
  File "/home/ubuntu/freqtrade/freqtrade/freqtradebot.py", line 1252, in update_trade_state
    if not isclose(order['amount'], new_amount, abs_tol=constants.MATH_CLOSE_PREC):
TypeError: must be real number, not NoneType
2020-08-17 14:14:23,684 - freqtrade.commands.trade_commands - INFO - worker found ... calling exit
2020-08-17 14:14:23,684 - freqtrade.rpc.rpc_manager - INFO - Sending rpc message: {'type': status, 'status': 'process died'}
2020-08-17 14:14:23,924 - freqtrade.freqtradebot - INFO - Cleaning up modules ...

**Comment by xmatthias** [comment]
@datanerdscott I expect this is fixed in #3597 .

If it is not, feel free to raise a new issue - but only if you run into this with an uptodate develop version.

**Comment by datanerdscott** [comment]
Thanks @xmatthias! Apologies, didn't realize it hadn't made it into master yet.  I have switched over to develop for my bittrex bots and will let you know if I encounter it again.  Thanks for the quick response!

**Comment by adrubesh** [comment]
@xmatthias  I'd like to reopen this as I am still encountering this error quite frequently in fact.
![image](https://user-images.githubusercontent.com/33096067/92995266-add04c00-f4c7-11ea-92ec-09c4a3969f7c.png)

I am in fact running develop as well.

**Comment by xmatthias** [comment]
@adrubesh best open a new issue, filling out the issue template, and providing the errorlog as text.
Also, please provide "more" of the log - not just the stacktrace - while it can be conclusive - most often errors come up before this, and the stacktrace is just showing the symptom - not the real reason.
